### PR TITLE
CIV-6804 filtering cases

### PR DESCRIPTION
**Pull request by dtortolaV1**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/CIV-6804

### Change description ###
Ensuring that the same event is not triggered twice for the same claim in the same polling iteration

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by sabahirfan — DISMISSED** [review]

**Review by indu2019 — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java [hunk 1]
```diff
@@ -130,4 +132,25 @@ private BusinessProcess businessProcessWithCamundaEvent(String camundaEvent) {
...
+                                                 a.getBusinessProcess().getCamundaEvent().equals(
+                                                     ((BusinessProcess)c.getData().get(
+                                                         "businessProcess")).getCamundaEvent())
+                                                     && a.getCcdCaseReference().equals(
+                                                     c.getId())
+                    ),
+                    ArgumentMatchers.eq(true)
+                )
+        );
+        verifyNoMoreInteractions(eventEmitterService);
```
Comment (filippo-possenti-version1) [inline src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java:154]:
```suggestion
        // And: only once for each case and camunda event
        verifyNoMoreInteractions(eventEmitterService);
```

File: src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java [hunk 2]
```diff
@@ -130,4 +132,25 @@ private BusinessProcess businessProcessWithCamundaEvent(String camundaEvent) {
             .status(READY)
             .build();
     }
+
+    @Test
+    public void doesNotRepeatEventClaim() {
```
Comment (filippo-possenti-version1) [inline src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java:137]: This test is not comprehensive.

It tests that you don't call Camunda twice for the same case with the same event, but what about the same case but a different event? That should probably be called (if I'm reading correctly your intentions) but you're not verifying it in the test.

Add to the `@BeforeEach` method a `caseDetails3b`, with the same `id` but a different `businessProcessWithCamundaEvent`. Then add it to the list of cases you are setting up on line 138 and 141.

**Review by filippo-possenti-version1 — COMMENTED** [review]
I would recommend to make the test more comprehensive, as I don't think it's capturing your full expectations of how the method should behave.

File: src/main/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandler.java [hunk 3]
```diff
@@ -28,13 +31,27 @@ public class PollingEventEmitterHandler implements BaseExternalTaskHandler {
...
-                eventEmitterService.emitBusinessProcessCamundaEvent(mappedCase, true);
+                if (caseEvent.add(Pair.of(
+                    mappedCase.getCcdCaseReference(),
+                    mappedCase.getBusinessProcess().getCamundaEvent()
+                ))) {
+                    log.info(format(
+                        "Emitting %s camunda event for case through poller: %d",
+                        mappedCase.getBusinessProcess().getCamundaEvent(),
+                        mappedCase.getCcdCaseReference()
+                    ));
```
[the remark this answers is not in the document]
Comment (dtortolaV1) [reply src/main/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandler.java:46]: Done

File: src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java (hunk 2 above)
Comment (dtortolaV1) [reply src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java:137]: Added in the new test not to affect other tests

File: src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java [hunk 4]
```diff
@@ -130,4 +132,25 @@ private BusinessProcess businessProcessWithCamundaEvent(String camundaEvent) {
...
+        when(searchService.getCases()).thenReturn(List.of(caseDetails1, caseDetails2, caseDetails3, caseDetails3));
```
[the remark this answers is not in the document]
Comment (dtortolaV1) [reply src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java:138]: done

File: src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java [hunk 5]
```diff
@@ -130,4 +132,25 @@ private BusinessProcess businessProcessWithCamundaEvent(String camundaEvent) {
...
+        pollingEventEmitterHandler.execute(externalTask, externalTaskService);
+
+        Stream.of(caseDetails1, caseDetails2, caseDetails3).forEach(
```
[the remark this answers is not in the document]
Comment (dtortolaV1) [reply src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java:141]: done

File: src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java (hunk 1 above)
Comment (dtortolaV1) [reply src/test/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandlerTest.java:154]: done

**Review by filippo-possenti-version1 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandler.java [hunk 6]
```diff
@@ -10,7 +11,9 @@
 import uk.gov.hmcts.reform.civil.service.EventEmitterService;
 import uk.gov.hmcts.reform.civil.service.search.CaseReadyBusinessProcessSearchService;
 
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 import static java.lang.String.format;
```
Comment (filippo-possenti-version1) [inline src/main/java/uk/gov/hmcts/reform/civil/handler/tasks/PollingEventEmitterHandler.java:18]: Checkstyle: unused import.

``suggestion
```

**Review by sabahirfan — APPROVED** [review]

**Pull request closed without merging**